src/coreclr/src/binder/inc/bindertracing.h

```diff
@@ -46,9 +53,128 @@ namespace BinderTracing
+    //
+    // As the resolution progresses to different stages, the GoToStage() method should be called.
+    // Calling it will fire an event for the previous stage with whatever context the class had
+    // at that point; it is assumed that if GoToStage() is called, the previous stage failed
+    // (the HRESULT is read by the dtor to assess success).
+    //
+    // It holds a reference to a HRESULT (that must be live during the lifetime of this object),
+    // which is used to determine the success or failure of a stage either at the moment this
+    // class is destructed (e.g. last stage), or when moving from one stage to another.  (This
+    // is especially useful if the HRESULT is captured from an exception handler.)
+    class ResolutionAttemptedOperation
+    {
+    public:
+        // This must match the ResolutionAttemptedStage value map in ClrEtwAll.man
+        enum class Stage : uint16_t
+        {
+            FindInLoadContext = 0,
+            AssemblyLoadContextLoad = 1,
+            PlatformAssemblies = 2,
+            DefaultAssemblyLoadContextFallback = 3,
+            ResolveSatelliteAssembly = 4,
+            AssemblyLoadContextResolvingEvent = 5,
+            AppDomainAssemblyResolveEvent = 6,
+            NotYetStarted = 0xffff, // Used as flag to not fire event; not present in value map
+        };
+
+        struct AttemptInfo
+        {
+            BINDER_SPACE::AssemblyName *AssemblyNameObject;
+            UINT_PTR BinderID;
+            INT_PTR ManagedAssemblyLoadContext;
+
+            PathString AssemblyName;
+            SString AssemblyLoadContext;
+        };
+
+    public: // static
+        static void TraceAppDomainAssemblyResolve(AssemblySpec *spec, PEAssembly *resultAssembly, Exception *exception = nullptr);
+
+    public:
+        ResolutionAttemptedOperation(BINDER_SPACE::AssemblyName *assemblyName, INT_PTR managedALC, const HRESULT& hr);
+        ResolutionAttemptedOperation(BINDER_SPACE::AssemblyName *assemblyName, UINT_PTR binderId, const HRESULT& hr);
+
+        void TraceBindResult(const BINDER_SPACE::BindResult &bindResult);
+
+        void SetFoundAssembly(BINDER_SPACE::Assembly *assembly)
+        {
+            m_pFoundAssembly = assembly;
+        }
+
+        void GoToStage(Stage stage)
+        {
+            assert(m_stage != stage);
+            assert(stage != Stage::NotYetStarted);
+
+            // Going to a different stage should only happen if the current
+            // stage failed (or if the binding process wasn't yet started).
+            // Firing the event at this point not only helps timing each binding
+            // stage, but avoids keeping track of which stages were reached to
+            // resolve the assembly.
+            TraceStage(m_stage, m_hr, m_pFoundAssembly);
+            m_stage = stage;
+            m_exceptionMessage.Clear();
+        }
+
+        void SetException(Exception *ex)
+        {
+            ex->GetMessage(m_exceptionMessage);
```

Comment:
I would think this should not do any work if tracing is not enabled (this makes a string copy - so not necessarily a cheap operation).